### PR TITLE
BUG: Rolling apply on DataFrame with Datetime index returns NaN

### LINKED ISSUE #15305: Rolling apply on DataFrame with Datetime index returns NaN
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame({'A': [0, 1, 2, 3, 4]},
                  index=[pd.Timestamp('20130101 09:00:00'),
                         pd.Timestamp('20130101 09:00:01'),
                         pd.Timestamp('20130101 09:00:02'),
                         pd.Timestamp('20130101 09:00:03'),
                         pd.Timestamp('20130101 09:00:04')])
rolled = df.rolling('3s', 3)

print rolled.sum()
print rolled.apply(sum)
```

outputs
```
                       A
2013-01-01 09:00:00  NaN
2013-01-01 09:00:01  NaN
2013-01-01 09:00:02  3.0
2013-01-01 09:00:03  6.0
2013-01-01 09:00:04  9.0

                      A
2013-01-01 09:00:00 NaN
2013-01-01 09:00:01 NaN
2013-01-01 09:00:02 NaN
2013-01-01 09:00:03 NaN
2013-01-01 09:00:04 NaN
```
#### Problem description
rolled.apply.sum() should give the same output as rolled.apply(sum) but instead only returns nan.
 

#### Expected Output
same as rolled.sum()

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-59-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 20.7.0
Cython: None
numpy: 1.12.0
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.1.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.5.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
boto: 2.38.0
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by FXocena**
- [x]  #15305
 - [x] changing the order of 'calling roll_sum' and 'get_window_indexer' in 'roll_generic' could fix #15305 
The bug caused by the return values from get_window_indexer. It changes value of 'win', which is not supposed to be passed to roll_sum.

### CODE REVIEW

**Comment by gfyoung** [comment]
> could fix

You can write a test to confirm your suspicion :wink:

**Comment by FXocena** [comment]
@gfyoung I add a test method to class TestRolling in test_window. 
The method is called 'test_constructor_with_timedelta_window_and_minperiods', using same dataframe as 'test_constructor_with_timedelta_window'.
with old window.pyx, this test will fail since the result_roll_generic is a all-nan dataframe

**Comment by gfyoung** [comment]
@FXocena : This looks pretty good.  You will need to address some style-checking errors <a href="https://travis-ci.org/pandas-dev/pandas/jobs/260999962#L1847">here</a> and add a `whatsnew` entry.  Otherwise, it looks great!

File: pandas/tests/test_window.py [hunk 1]
```diff
@@ -423,6 +423,27 @@ def test_constructor_with_timedelta_window(self):
...
+        df = pd.DataFrame({'value': np.arange(n)},
+                          index=pd.date_range('2015-12-24',
+                                              periods=n,
+                                              freq="D"))
+        expected_data = np.append([np.NaN, 1.], np.arange(3., 27., 3))
+        for window in [timedelta(days=3), pd.Timedelta(days=3)]:
+            result_roll_sum = df.rolling(window=window, min_periods=2).sum()
+            result_roll_generic = df.rolling(window=window,
+                                             min_periods=2).apply(sum)
+            expected = pd.DataFrame({'value': expected_data},
```
Comment (jreback) [inline pandas/tests/test_window.py]: you can put the expected_data directly in the frame.

File: pandas/tests/test_window.py [hunk 2]
```diff
@@ -423,6 +423,27 @@ def test_constructor_with_timedelta_window(self):
...
+                                    index=pd.date_range('2015-12-24',
+                                                        periods=n,
+                                                        freq="D"))
+            tm.assert_frame_equal(result_roll_sum, expected)
+            tm.assert_frame_equal(result_roll_generic, expected)
+            expected = df.rolling('3D', min_periods=2).sum()
```
Comment (jreback) [inline pandas/tests/test_window.py]: what is this last part testing?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/test_window.py (hunk 2 above)
Comment (FXocena) [inline pandas/tests/test_window.py]: This tests when window == '3D'. I'd better put it together with [timedelta(days=3), pd.Timedelta(days=3)]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
ok pushed a commit. this can be merged on green.

**Comment by jreback** [comment]
thanks!

**Pull request merged**